glusterd2/pmap/registry.go

```diff
@@ -0,0 +1,226 @@
+					// a brick process
+					break
+				} else {
+					r.ports[p].State = PortForeign
+					r.PortUnlock(p)
+				}
+			}
+		}
+	}
+
+	if port == 0 {
+		return -1, fmt.Errorf("registry.Allocate(): we ran out of free ports")
+	}
+
+	if port > r.lastAlloc {
+		r.lastAlloc = port
+	}
+
+	return port, nil
+}
+
+// Update marks the port used by a brick with a specified state. This
+// is called when a brick signs in.
+func (r *pmapRegistry) Update(port int, brickpath string, conn net.Conn) error {
+
+	if port < 0 || port > PortMax {
+		return fmt.Errorf("registry.Update(): invalid port %d", port)
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	// It's possible that multiple bricks are multiplexed onto a
+	// single conn, the conn passed to this function may not be the
+	// same as before and that can happen. We only store the latest
+	// one.
+	r.conns[conn] = port
+	r.ports[port].State = PortInUse
+	if r.ports[port].Bricks == nil {
+		// can happen on glusterd2 restarts where we only get a SIGN IN
+		r.ports[port].Bricks = make(map[string]bool)
+	}
+	r.ports[port].Bricks[brickpath] = true
+	r.bricks[brickpath] = port
+
+	if r.lastAlloc < port {
+		r.lastAlloc = port
+	}
+
+	return nil
+}
+
+// SearchByBrickPath returns the port number used by the brick specified
+// by the brick path provided.
+func (r *pmapRegistry) SearchByBrickPath(brickpath string) (int, error) {
+
+	if brickpath == "" {
+		return -1, fmt.Errorf("SearchByBrickPath: brick path cannot be empty")
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+
+	if port, ok := r.bricks[brickpath]; ok {
+		return port, nil
+	}
+
+	return -1, fmt.Errorf("SearchByBrickPath: port for brick %s not found", brickpath)
+}
+
+// RemoveByConn deletes port map entry by brick process's TCP connection.
+// There will be only one TCP connection per brick process, regardless of
+// number of bricks in the process.
+func (r *pmapRegistry) RemovePortByConn(conn net.Conn) error {
+
+	if conn == nil {
+		return fmt.Errorf("RemovePortByConn(): conn passed is nil")
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	port, ok := r.conns[conn]
+	if !ok {
+		// this can happen in many cases:
+		// * conn isn't a brick
+		// * brick disconnects prior to SIGN IN
+		return nil
+	}
+
+	delete(r.conns, conn)
+
+	r.ports[port].State = PortFree
+	r.PortUnlock(port)
+
+	for brick := range r.ports[port].Bricks {
+		delete(r.bricks, brick)
+	}
+	r.ports[port].Bricks = make(map[string]bool)
+
+	return nil
+}
+
+// Remove deletes portmap entry of a single brick from the portmap registry.
+// This is called when a brick process sends a SIGN OUT request to glusterd2
+// during graceful shutdown.
+func (r *pmapRegistry) Remove(port int, brickpath string, conn net.Conn) error {
+
+	if port < 0 || port > PortMax {
+		return fmt.Errorf("registry.Remove(): invalid port %d", port)
+	}
+
+	r.Lock()
+	defer r.Unlock()
+
+	delete(r.bricks, brickpath)
+	if _, ok := r.ports[port].Bricks[brickpath]; ok {
+		delete(r.ports[port].Bricks, brickpath)
```

Comment:
If it was the last brick being removed, shouldn't you be resetting the port?

Reply:
The last brick being removed will eventually disconnect, at which point, `RemovePortByConn` (see line 171 and 172) will reset the port.